[1단계 - 자동차 경주 구현] 렉스(오성원) 미션 제출합니다.

Seongwon97 opened the pull request:
안녕하세요! 😁
저는 렉스라고 합니다! (토이스토리의 공룡 장난감 이름을 따왔어요🦖)
먼저 바쁘실텐데 시간을 내주신 점에 감사드리고 싶습니다.

이번 미션은 이전 프리코스 때 진행하였던 미션임에도 불구하고 짝과 함께 페어 프로그래밍을 진행하고 처음 해보는 TDD 방법론에 따라 미션을 진행하다보니 많은 시간을 투자했던 것 같습니다.

TDD 방법론으로의 개발은 프리코스 과정때도 시도를 해보고 싶었으나 이론상으로는 이해를 하여도 막상 코드로 구현은 어떠한 방법으로 진행을 해야하는지 몰라서 하지 못하였는데 우테코의 교육과 미션을 통해 어느정도 감을 잡은것 같습니다.

페어 프로그래밍은 처음에 짝과 함께 개발을 하다보니 혼자 개발을 하는 것보다 많은 시간이 소모되어 이게 좋은 개발 방법인가?하는 의문감이 많이 들곤 하였는데 페어와 직접 만나서 페어 프로그래밍을 진행해보니 서로의 부족한 부분을 채울 수 있었고, 서로의 코딩 스타일과 여러 의견들을 조율해가며 개발하다보니 새로운 경험을 한 것 같습니다.

그리고 마지막에 페어의 깃으로부터 저의 깃으로 commit Log와 코드를 가져오는 단계에서 그동안 사용해보지 못한 깃의 여러 기능을 접하게 되었고 깃을 제대로 공부해봐야겠다는 다짐을 하게 되었습니다.

아직 배워야 할 것이 많은 개발자이지만 앞으로의 강의와 피드백들을 통해 발전하는 개발자가 되도록 하겠습니다🔥

### 미션을 진행하며 궁금하였던 점은 다음과 같습니다.

- 생성자에서 값을 초기화 할때도 상수를 사용하는 것이 좋나요?
Car객체에 private static final int INITIAL_POSITION = 0;와 같이 자동차 위치를 0이라는 수로 초기화를 할 때 상수로 초기화하지 않아도 코드를 보면 해당 코드가 자동차의 초기 위치를 설정하는 것을 알 수 있는데 이러한 값들도 상수를 지정해주는게 좋은지 알고 싶습니다. (상수 정의의 기준을 알고 싶습니다.)
- 이 외에도 Test와 미션에 적용하는 mvc에 관해 궁금하였던 점이 많았으나 오늘 오전 네오님의 강의와 데일리 미팅에서 구구님의 피드백을 들으며 많은 궁금증들이 해소되었습니다.

<hr>

감사합니다 🙇🏻‍♂️

skullkim committed:
study(test/java/study): 학습테스트를 통해 Junit 학습하기 완료

String API, Set collection을 이용해 Junit5 연습 완료

skullkim committed:
study(StringCalculator.java): 문자열 덧셈 계산기 완료

문자열 덧셈 계산기구현을 통해 TDD공부

skullkim committed:
feat(RandomNumber.java): 난수생성기능 완료

자동차 전진 여부를 판단하기 위해 필요한 난수를 생성하는 기능 완료

skullkim committed:
feat(Car.java):자동차 한대의 정보를 담는 클래스 생성

랜덤 메서드를 호출하여 전진 여부를 판단 후 전진하는 메서드 랜덤 메서드를

skullkim committed:
refactor(RandomNumber.java):파일 위치 조정

RandomNumber.java파일을 /utils 내부로 이동

BREKAING CHANGE: RandomNumner.java 파일 위치 수정
as-is:
main/java/racingcar/RandomNumber.java
to-be:
main/java/racingcar/utils/RandomNumber.java

skullkim committed:
feat(InputValidation.java):입력된 데이터 검증 기능 구현

입력한 이름들의 길이가 각 5가 넘지 않도록 검증

skullkim committed:
test(InputValidationTest.java): 테스트 추가

이름 데이터를 검증하는 기능에 대한 테스트 추가

유효하지 않은 이름(6자 이상)이 왔을 때 예외를 던지는지 확인하는 테스트
추가

skullkim committed:
feat(InputValidation.java): 입력된 반복 횟수 검증

입력한 반복 횟수가 정수가 하닌 다른 문자가 포함되 있다면
RunTimeException을 던진다.

skullkim committed:
feat(main/java/views): 입출력 기능 완료

게임 진행을 위해 사용자가 입력해야 하는 자동차 이름과 턴 수를 입력받는
기능 완료. 게임 진행 과정과 결과를 알려주는 기능 완료.
입력이 유효하지 않으면 유효한 입력이 있을 떄 까지 다시 입력받는다.

skullkim committed:
feat(RacingGame.java): 레이싱 게임 로직 구현

이전에 구현한 기능을 활용해 실제 레이싱 게임 로직 구현

skullkim committed:
refactor(src/main/java/racingcar): 불변이어야 하는 변수에 final 추가

불변이여야 하는 지역 변수와 파리미터에 final 추가

skullkim committed:
refactor(Input.java): 에러 문구 출력 기능 추가

잘못된 입력이 있을 때 에러를 알려주는 문구를 출력하는 기능 추가.

BREAK CHANGE: Input class의 inputNames, inputRepeats 메서드 변경
inputNames 메서드:
as-is:
private String inputNames() {
    try {
        ...
    } catch (RuntimeException exception) {
        return inputNames();
    }
}

to-be:
private String inputNames() {
    try {
        ...
    } catch (RuntimeException exception) {
        System.out.println(exception.getMessage() + "\n");
        return inputNames();
    }
}

inputRepeats 메서드:
as-is:
private String inputRepeats() {
    try {
        ...
    } catch (RuntimeException exception) {
        return inputRepeats();
    }
}

to-be:
private String inputRepeats() {
    try {
        ...
    } catch (RuntimeException exception) {
        System.out.println(exception.getMessage() + "\n");
        return inputRepeats();
    }
}

skullkim committed:
refactor(src/main/java.racingcar): 클래스 컨벤션 수정

클래스 시작 부분을 하나의 공백라인으로 시작하게 수정.

skullkim committed:
refactor(src/main/java/racingcar): 게임에 필요한 메시지 통합

게임 진행을 위한 입출력, 에러 안내 메시지를 하나의 클래스에서 관리하기
위해 RacingGameMessage 생성. RacingGameMessage에서 모든 메시지를
관리한다.

skullkim committed:
refactor(Car.java): 상수 추가

전진 여부를 판단하는 숫자와 차 위치 초기값을 상수로 변경.

FORWARD_STANDARD = 4(자동차 전진 여부)
INITIAL_POSITION = 0(초기 차량 위치값)

skullkim committed:
Merge branch 'racing-game' into develop

ddu0422 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,49 @@
+import java.util.regex.Pattern;
+
+public class StringCalculator {
+	private static final String CUSTOM_DISTRIBUTOR_PATTERN = "//(.)\n(.*)";
```
해당 클래스만 중괄호와 상수 사이에 공백이 없는 이유가 있을까요?

Seongwon97 replied:
Calculator를 먼저 구현한 후 Racingcar을 구현하며 RacingCar의 코드 통일에만 신경쓰다보니 놓친 것 같습니다.🥲 
바로 수정하였습니다!

ddu0422 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,49 @@
+	private static final String BASIC_DISTRIBUTOR_PATTERN = ",|:";
+	private static final String NUMBER_PATTERN = "^[0-9]*$";
+	private static final String INVALID_INPUT_ERROR = "정수가 아닌 문자가 포함되어 있습니다.";
+	private static final int NUMBERS = 2;
```
NUMBERS보단 의미 있는 네이밍으로 변경해보는 건 어떨까요?

Seongwon97 replied:
NUMBERS가 아닌 더욱 의미있어보이는 그룹 내의 NUMBER라는 의미를 내고자하여 NUMBER_GROUP으로 수정하였습니다❗️

ddu0422 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,49 @@
+			return 0;
+		}
+		Matcher matcher = Pattern.compile(CUSTOM_DISTRIBUTOR_PATTERN).matcher(input);
+		if (matcher.find()) {
```
해당 메서드가 실행될 때마다 Pattern 객체가 계속 생성되겠네요 :)
값이 변하지 않으니 static으로 선언하여 불필요한 생성을 줄여보아요.

Seongwon97 replied:
피드백 주신대로 Pattern객체를 static Import하도록 코드를 수정하니 불필요한 생성을 줄인것을 넘어 코드도 더 깔끔해보니는 것 같네요😆

ddu0422 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,49 @@
+
+	private static void checkInvalidInput(String input, String distributor) {
+		if (!Pattern.matches(NUMBER_PATTERN,
+				String.join("", splitNumber(input, distributor)))) {
```
모니터의 크기가 작아 예전엔 라인 수당 글자 수가 80자가 Convention이긴 했지만, 현재는 모니터의 크기가 커졌기 때문에 120자까지 하는 경우가 많습니다.
다른 라인도 80자가 넘어가는 경우가 많으니 120자로 통일하는 건 어떨까요?

ddu0422 commented on src/main/java/calculator/StringCalculator.java:
```diff
@@ -0,0 +1,49 @@
+	private static Integer getSum(String input, String distributor) {
+		return Arrays.stream(splitNumber(input, distributor))
+			.map(Integer::parseInt)
+			.reduce(0, Integer::sum);
```
저는 `mapToInt().sum()`을 사용하면 가독성이 더 좋더라구요 :)

Seongwon97 replied:
코드를 `mapToInt().sum()` 으로 변경을 하다보니 제가 변경한 방법이 옳바르게 바꾼 방법인지 궁금해서 질문을 남기게 되었습니다. Stream.java파일에서 `mapToInt`를 찾아보니 파라미터로 integer로 변경할 mapper를 넘겨줘야 하더군요. 구글에서 mapToInt 예시를 찾아봤을 때 다른 사람들의 코드들은 Arrays스트림이 아닌 List<클래스A>타입의 값들을 stream으로 사용을하며 mapToInt의 mapper로는 클래스A안에 만들어준 매서드가 들어가더라고요.. 저는 String의 값을 단순히 int타입으로만 바꾸도록 설정을 하였는데 옳바르게 작성한 것일까요?? 
`number->Integer.parseInt(number)`부분을 더욱 간략하게 표현할 방법은 따로 없을까요?
```java
// 변경한 코드입니다.
return Arrays.stream(splitNumber(input, distributor))
        .mapToInt(number -> Integer.parseInt(number))
        .sum();
```

ddu0422 replied:
잘 작성하셨어요. 올바르게 작성했다고 생각해요 :)
`number -> Integer.parseInt(number)` 부분은 메서드 레퍼런스 부분을 찾아보시면 될 것 같습니다 !_!

추가로 아래 내용을 앞으로 미션을 진행하면서 찾아보시면 도움이 될 것 같아요 😉

1. Arrays에서 List<T> 에서 사용하는 stream을 어떻게 사용하고 있는 지
2. .map()과 .mapToInt() / .mapToDouble() 등은 어떤 차이가 있는 지

ddu0422 commented on src/main/java/racingcar/controller/RacingGame.java:
```diff
@@ -0,0 +1,76 @@
+	final Input input;
+	final Output output;
+	ArrayList<Car> cars;
+	int repeats;
```
해당 클래스에서만 사용하는데 package-private 지시자를 사용한 이유가 있을까요?

Seongwon97 replied:
실수로 놓친것 같습니다...수정하였습니다‼️

ddu0422 commented on src/main/java/racingcar/controller/RacingGame.java:
```diff
@@ -0,0 +1,76 @@
+
+	final Input input;
+	final Output output;
+	ArrayList<Car> cars;
```
ArrayList를 사용하는 경우와 List를 사용하는 경우 어떤 차이가 있을까요?

Seongwon97 replied:
이전에 학습을 할 때 `ArrayList<Car> cars` 가 아닌 `List<Car> car` 와 같이 사용한다면 범용성이 늘어난다고 학습을 하였던 기억이 있습니다. 개발을 하다보면 처음 생각하였던 것과 다르게 ArrayList가 아닌 LinkedList 와 같은 다른 List들을 사용하게 될 수도 있고 추후에 유지보수를 하며 변경을 하는 상황이 발생할 수 있는데 이러한 상황을 List로 사용할 경우 더 좋다고 기억하고있습니다.
제가 알고있는 내용들이 맞을까요??

ddu0422 replied:
네네 맞습니다 :)
렉스가 이야기한대로 ArrayList에서 LinkedList 변환은 유지보수 측면에서 그나마 간단하겠지만 점점 시스템이 커질수록 구현체에 의존하게 된다면 유지보수 측면에서 힘들어지게 됩니다 !_!..

[해당 내용](https://hirlawldo.tistory.com/113)도 읽어보면 좋을 것 같아요

ddu0422 commented on src/main/java/racingcar/controller/RacingGame.java:
```diff
@@ -0,0 +1,76 @@
+	}
+
+	private void createCar() {
+		final List<String> carNameList = input.inputValidNames();
```
변수명에 자료형을 함께 쓰면 나중에 자료형이 변경되는 경우 변수명도 변경되어야 해요.
개발자의 실수로 자료형 변경 후 변수명을 변경하지 않는 상황이 발생한다면 나중에 해당 코드의 의미를 오해하는 상황이 발생해요.
저는 변수명에 Collection 자료형을 사용하는 대신 복수 변수명으로 선언하는 편입니다 :)
findWinner 함수에 winners라는 좋은 예제도 있군요 !_!

Seongwon97 replied:
변수명에는 자료형을 사용하기보다는 복수 변수명과 같이 자료형이 들어가지 않는 이름을 사용하자!!
덕분에 몰랐던 내용을 하나 더 알아갑니다😁

ddu0422 commented on src/main/java/racingcar/controller/RacingGame.java:
```diff
@@ -0,0 +1,76 @@
+		return cars.stream()
+			.sorted(Comparator.comparing(Car::getPosition))
+			.collect(Collectors.toList())
+			.get(cars.size() - 1).getPosition();
```
해당 Applicadtion에선 cars.size()가 0일 가능성은 없지만 누군가 해당 메서드를 사용할 때 0인 경우 에러가 발생할 가능성이 있지 않을까요?

Seongwon97 replied:
피드백을 반영하여 결과로 나오는 List가  isEmpty()가 아닐 경우에만 List첫번째 값을 return 하도록 수정하였습니다 !!

ddu0422 commented on src/main/java/racingcar/controller/RacingGame.java:
```diff
@@ -0,0 +1,76 @@
+
+	private List<Car> getWinners(final int farthestPosition) {
+		return cars.stream()
+			.filter((car) -> farthestPosition == car.getPosition())
```
자동차의 위치를 가져와서 비교하는 것이 아닌 메시지를 보내는 방식으로 변경해보면 어떨까요?

Seongwon97 replied:
피드백을 반영하여 joinWinnerNames()메서드의 작업을 getWinners()메서드에서 한번에 동작하여 메시지를 보내는 방식으로 변경해봤습니다!!

ddu0422 replied:
앗..! 해당 방법은 메서드 분리라고 생각해요 :) Car 객체에 `메시지를 전달`하라는 의미였어요 !_!
정확한 부분은 `car.getPosition()` 부분이 문제라고 생각해서 리뷰를 남겼어요.

나중 미션에 배우겠지만 많이 고민하신 것 같아 살짝 힌트를 드리자면

메시지를 전달하라는 말은 그 객체에게 물어보라는 의미예요.
렉스가 저에게 질문을 하듯 RacingGame이 Car 객체에게 질문을 해보면 어떨까요?

ddu0422 commented on src/main/java/racingcar/controller/RacingGame.java:
```diff
@@ -0,0 +1,76 @@
+		final String[] winnerNames = winners.stream()
+			.map(Car::getName)
+			.toArray(String[]::new);
+		return String.join(", ", winnerNames);
```
`.collect(Collectors.joining(", "));` 을 사용하는 방식도 있습니다 😄

Seongwon97 replied:
Collectors에 joining기능도 있었군요😁  해당 기능을 몰라 List를 String배열로 변경하고 String.join()을 사용했었는데 미르님이 알려주신 `.collect(Collectors.joining(", "));`를 사용하니 코드가 훨씬 간결해진 것 같습니다!! 
감사합니다😁

ddu0422 commented on src/main/java/racingcar/models/Car.java:
```diff
@@ -0,0 +1,33 @@
+
+	public int getPosition() {
+		return position;
+	}
+}
```
equals && hashcode & toString을 재정의해보는 건 어떨까요?
재정의하는 이유는 무엇인지 고민해보면 좋을 것 같아요 :)

Seongwon97 replied:
equals && hashcode & toString 을 모두 재정의해보았습니다.

하지만 toString의 경우는 결과 출력을 할 때 해당 객체 자체를 이용하여 출력을 할 수 있다는 이점을 느꼈으나 equals, hashCode의 경우는 해당 프로그래밍에서 객체 비교를 할 일이 없었어서 실제 사용되지는 못한 것 같습니다.

해당 매서드들 재정이의 장점은 사용자가 직접 equals, hashCode들을 재정의하며 사용자가 정한 규칙으로 객체 비교를 할 수 있으며 toString을 통해 사용자가 원하는 방법으로 해당 객체의 출력값을 정할 수 있다는 점이라 생각합니다.

ddu0422 replied:
렉스의 의견대로 equals와 hashCode는 현재 상황에선 불필요해요 💯 💯 
한번 정의해보면 어떤 점이 좋을 지 생각해보라는 의미였어요 ㅎ_ㅎ
미션을 진행하다보면 점차 알게 되리라고 생각해요! 

하지만 현재 toString을 getName과 동일하게 정의하셨군요! 처음에는 UI와 Domain이 분리되어 있었는데
toString을 추가하고 오히려 UI와 Domain이 엮여있는 형태가 되었어요 ㅜㅡㅜ..
UI가 변경이 된다면 toString이 변경되어야할까요 ?_? 

toString을 정의하지 않고 생성한 Car 객체를 출력했을 때와 정의하고 생성한 Car 객체를 출력하면 어떤 값이 나올까요?
toString을 재정의 하는 목적은 https://lelecoder.com/42 를 참고하면 좋을 것 같아요 :)

ddu0422 commented on src/main/java/racingcar/utils/RacingGameMessage.java:
```diff
@@ -0,0 +1,17 @@
+	public static final String CAR_POSITION = "-";
+	public static final String NAME_POSITION_SPLITTER = " : ";
+
+	public final static String CAR_NAME_DISTRIBUTOR = ",";
```
상수도 각 책임에 맞게 각 클래스에 정의하는 건 어떨까요?

Seongwon97 replied:
피드백을 반영하여 상수들을 책임을 갖는 각 클래스에 정의하도록 변경해봤습니다.

프리코스를 진행할 때 부터 상수들을 따로 파일을 생성하여 관리를 하는 것이 좋은지 책임을 갖는 클래스 내에 정의를 하는것이 좋을지 궁금했는데 미르님은 각각의 클래스에 정의하는 것을 추천하시나요??

만약 Input.java파일과 InputValidation.java파일에서 같이 사용하는 `CAR_NAME_DISTRIBUTOR`같은 것은 한 쪽에서 public으로 정의를 하고 다른 한쪽에서 Import를 사용하는 것이 좋을까요..? 아니면 양쪽에 private으로 각각 정의를 하여 사용하는 것이 좋을까요...?? 실무에서는 주로 어떠한 방법을 사용하는지 궁금합니다!!🧐

ddu0422 replied:
실무에서는 정말 케이스가 다양해요. 공통으로 쓰고 있는 팀도 있고 각 클래스마다 정의하는 팀도 있어요.
그러다 보니 공통으로 쓰고 있는 팀의 코드를 볼 때는 상수 길이가 점점 길어져서 해당 파일의 라인 수가 1000 ~ 2000 줄 이렇게 되어 있더라구요?
해당 파일 내에서도 동일한 값이지만 다른 네이밍으로 되어 있는 경우도 많이 있었어요.
그래서 저는 각각의 책임이 있다고 생각해서 별도 클래스 두기보다는 관련 클래스에서 사용할 것 같아요 :) CAR_NAME_DISTRIBUTOR처럼 공통으로 쓰일 수 있지만 변경이 자주 일어나지 않다면 별도로 각 클래스에 맞게 정의할 것 같아요 !_!

고민을 해봐도 다른 클래스로 분리하는 게 의미가 더 크다고 생각하면 다른 클래스로 두셔도 됩니다 !!
단순히 제 생각이니 렉스도 고민하며 개발 철학을 잡아가면 좋을 것 같아요 :)

ddu0422 commented on src/main/java/racingcar/utils/RacingGameMessage.java:
```diff
@@ -0,0 +1,17 @@
+	public static final String INPUT_REPEATS = "시도할 회수는 몇회인가요?";
+
+	public static final String NAME_LENGTH_ERROR = "이름은 5자리 이하만 가능합니다.";
+	public static final String REPEATS_ERROR = "반복횟수는 정수만 입력 가능합니다.";
```
음의 정수를 입력해도 해당 메시지가 나오는데 다른 에러 메시지가 어울리지 않을까요?

Seongwon97 replied:
Input.java파일의 inputRepeats메서드에 조건을 추가하여 상황에 따라 다른 메시지가 출력되도록 변경해봤습니다!

ddu0422 commented on src/main/java/racingcar/utils/RandomNumber.java:
```diff
@@ -0,0 +1,13 @@
+	private final static int MAX_EXCLUSIVE = 9;
+
+	public int getRandomNumber() {
+		final Random random = new Random();
```
random을 계속 생성하기보단 static으로 선언하여 재사용하는 건 어떨까요?

Seongwon97 replied:
피드백을 반영하여 Random객체 생성을 반복적으로 하던 코드를 재사용하는 코드로 변경하였습니다!

ddu0422 commented on src/main/java/racingcar/utils/RandomNumber.java:
```diff
@@ -0,0 +1,13 @@
+
+	public int getRandomNumber() {
+		final Random random = new Random();
+		return random.nextInt(MAX_EXCLUSIVE) + 1;
```
요구사항은 `전진하는 조건은 0에서 9 사이에서 random 값을 구한 후...` 이군요!

Seongwon97 replied:
앗...요구사항에 맞게 코드 수정하였습니다!!

ddu0422 commented on src/main/java/racingcar/views/Input.java:
```diff
@@ -0,0 +1,54 @@
+			inputValidation.validateName(names);
+			return names;
+		} catch (RuntimeException exception) {
+			System.out.println(exception.getMessage() + "\n");
```
\n보단 System.lineSeparator(); 을 사용하는 건 어떨까요? OS마다 개행이 달라질 수 있어요 :)

Seongwon97 replied:
`System.lineSeparator()`는 처음 접한 코드인 것 같습니다. 간단한 줄바꿈도 OS마다 개행이 달라질 수 있다니..새로운 사실을 알아가게 된 것 같습니다!

앞으로 줄바꿈에는 `System.lineSeparator()`를 사용해야겠군요!!
감사합니다

ddu0422 commented on src/main/java/racingcar/views/Output.java:
```diff
@@ -0,0 +1,30 @@
+		System.out.print("\n");
+	}
+
+	public void printWinner(final String winners) {
```
매개변수에 winners를 받고 있으니 printWinners는 어떨까요?

Seongwon97 replied:
복수의 우승자를 받고 한 사람의 우승자만 출력하는 것 처럼 메서드 명이 정의되어 확실히 통일감이 없었네요😅   printWinners로 수정해봤습니다!!

ddu0422 commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,47 @@
+
+public class StringCalculatorTest {
+	@Test
+	public void splitAndSum_null_또는_빈문자() throws Exception {
```
Exception은 불필요한 내용인 것 같아요 :)

Seongwon97 replied:
불필요한 Exception들 모두 제거하였습니당~

ddu0422 commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,47 @@
+	public void splitAndSum_custom_구분자() throws Exception {
+		int result = StringCalculator.splitAndSum("//;\n1;2;3");
+		assertThat(result).isEqualTo(6);
+	}
```
테스트가 실패하네요 😭

Seongwon97 replied:
CUSTOM_DISTRIBUTOR 상수값이 잘못 설정되어 있었네요ㅠ 수정하였습니다!

ddu0422 commented on src/test/java/racingcar/controller/RacingGameTest.java:
```diff
@@ -0,0 +1,27 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+@DisplayName("RacingGame 클래스")
```
Root Class에 DisplayName을 통일하는 건 어떨까요?

Seongwon97 replied:
class에 붙은 DisplayName은 모두 제각각으로 정의되어 있었네요...😅 
모두 같은 형식으로 통일시켰습니다!

ddu0422 commented on src/test/java/racingcar/controller/RacingGameTest.java:
```diff
@@ -0,0 +1,27 @@
+	@Test
+	@DisplayName("게임이 정상적으로 실행되는지 확인한다")
+	void runGame() {
+		final RacingGame racingGame = mock(RacingGame.class);
```
Mock Good !! 💯

ddu0422 commented on src/test/java/racingcar/models/CarTest.java:
```diff
@@ -0,0 +1,39 @@
+
+	@Test
+	@DisplayName("자동차가 앞으로 전진하는지 확인한다")
+	void goForward() {
+		when(randomNumber.getRandomNumber()).thenReturn(6);
+		Car car = new Car(randomNumber, "a");
+		int exPosition = car.getPosition();
+
+		car.goForward();
+
+		assertThat(exPosition + 1).isEqualTo(car.getPosition());
```
해당 내용은 Mock으로 구현하는 것도 좋지만 다른 방식으로 구현해보는 건 어떨까요?

Seongwon97 replied:
해당 테스트는 아무리 생각을 해봐도 Mock으로 테스트하는 방법 이외에 어떠한 방법으로 테스트를 진행해야할지 떠오르지가 않네요...

goForward()메서드가 RandomNumber객체의 getRandomNumber()메서드의 결과값에 따라 전진, 유지의 결과가 나와서 Mock이외에 테스트를 할 방법이 떠오르지 않습니다 🥲

혹시 다른 테스트 방법이 있다면 알려주실 수 있으실까요..?

ddu0422 replied:
고민을 많이 하셨군요 ㅜ_ㅜ...
아마 오늘 강의에서 해당 내용을 배울 것 같아요 😄 만약 강의 시간에 다루지 않았거나 해당 내용을 들어도 궁금한 점이 생기면 Comment나 DM 주세요!

그 전에 살짝 힌트를 드리자면 테스트하기 쉽지 않은 이유는 Random 값 때문인데요.
해당 메서드에는 변하는 부분과 변하지 않는 부분이 같이 포함되어 있어요. 그 둘을 분리하여 변하지 않는 부분만을 남기려면 어떤 방식이 필요할 지 고민해보면 좋을 것 같아요 :)

ddu0422 commented on src/test/java/racingcar/utils/InputValidationTest.java:
```diff
@@ -0,0 +1,62 @@
+		void checkRightName(final String names) {
+			InputValidation validation = mock(InputValidation.class);
+			validation.validateName(names);
+			verify(validation).validateName(names);
```
mocking을 하는 것 보단 `assertDoesNotThrow()`을 활용하여 에러나 나지 않는 지 확인하는 것도 좋을 것 같아요 :)
다음 미션들을 진행할 때 mocking을 사용하는 이유가 무엇인지 고민해보고 적용할 곳과 적용하지 않아야 할 곳을 고민해보면 좋을 것 같아요.

Seongwon97 replied:
피드백을 반영하여 테스트가 단순히 실행되는 것을 확인하는 `verify`를 `assertDoesNotThrow()`를 통해 에러가 발생하지 않는지 확인하도록 수정하였습니다!!

ddu0422 commented on src/test/java/study/SetTest.java:
```diff
@@ -0,0 +1,47 @@
+		numbers.add(3);
+	}
+
+	@Test
+	@DisplayName("Set 컬렉션의 크기를 구한다.")
+	void sizeTest() {
+		assertThat(numbers.size()).isEqualTo(3);
+	}
+
+	@DisplayName("Set 컬랫션에 원소가 존재하는지 확인한다")
+	@ParameterizedTest(name = "{index} {displayName} key={0}")
+	@ValueSource(ints = {1, 2, 3})
+	void containTest(final int key) {
+		assertThat(numbers.contains(key)).isTrue();
+	}
+
+	@DisplayName("Key 값으로 value값의 존재를 판단한다.")
+	@ParameterizedTest(name = "{index} {displayName} key={0} expected={1}")
+	@CsvSource(value = {"1:true","4:false"}, delimiter =':')
+	void findValueTest(int key, boolean expected) {
+		assertThat(numbers.contains(key)).isEqualTo(expected);
+	}
```
학습 테스트 💯

ddu0422 reviewed (changes requested):
안녕하세요 렉스! 리뷰어 미르입니다 :)
렉스라는 닉네임 부르기 되게 좋네요 ㅎ_ㅎ
Nested를 이용하여 테스트하신 점이 인상깊네요 😄 

TDD 처음 접할 때 정말 어렵죠 😭  저도 처음 접하고 시도했을 때 많이 어려웠어요
하지만 꾸준한 노력을 하신다면 금방 익숙해질 수 있을거예요 📈 

첫 구현 잘 마무리 해주셨네요 💯
제가 중점적으로 본 내용은 프로그램 동작 여부, 테스트 코드, 객체 지향 생활 체조, 읽기 좋은 코드입니다.
궁금한 점이 생기면 언제든지 DM 혹은 댓글 남겨주세요 ^-^

## Q&A
> 생성자에서 값을 초기화 할때도 상수를 사용하는 것이 좋나요?
Car객체에 private static final int INITIAL_POSITION = 0;와 같이 자동차 위치를 0이라는 수로 초기화를 할 때 상수로 초기화하지 않아도 코드를 보면 해당 코드가 자동차의 초기 위치를 설정하는 것을 알 수 있는데 이러한 값들도 상수를 지정해주는게 좋은지 알고 싶습니다. (상수 정의의 기준을 알고 싶습니다.)

상수를 선언하는 편이 좋다고 생각합니다 :) 상수를 정의하는 기준을 알면 자연스레 해결되겠네요.
상수를 정의하는 기준은 누구나 다를 수 있다는 점을 고려해주시면 감사하겠습니다 :)

제가 상수를 정의하는 기준은 상수에 이름을 부여하여 코드를 읽을 때 `이 숫자, 문자열은 무엇을 의미하는 거지?`라는 생각이 들지 않도록 하기 위함입니다.
코드 전체적인 맥락을 파악하면 0이 자동차의 초기 위치를 나타냄을 알 수 있지만, 해당 도메인을 모르는 경우라면 어떻게 될까요? 문제를 파악해야 해당 상수의 의미를 알 수 있지 않을까요? 그렇기에 코드를 파악하며 이해를 돕기 위해 상수로 선언합니다 !_!
해당 코드를 보는 누군가가 이해하기 쉽다고 생각하면 상수로 선언하지 않아도 될 수도 있구요 ㅎ_ㅎ

앞으로도 개발을 많이 하실 텐데 렉스만의 기준을 잡아보면 좋을 것 같아요 :)

좋은 질문 주셔서 감사합니다 렉스 🙇🏻‍♂️

Seongwon97 committed:
refactor(SpringCalculator.java) : 피드백을 반영하여 코드 수정

- SpringCalculator.java파일의 중괄호와 상수 사이 공백 추가
- NUMBERS 상수를 더 의미있는 이름인 NUMBER_GROUP으로 변경
- 반복적으로 생성되는 Pattern객체를 static으로 import추가
- 36번 라인의 줄 바꿈을 제거
- getSum메서드의 map().reduce()를 mapToInt().sum()으로 변경
- SpringCalculatorTest.java파일 내의 메서드에 붙어있는 불필요한 throw Exception제거
- splitAndSum_custom_구분자()메서드의 실패 오류 수정

Seongwon97 committed:
refactor(RacingGame.java): 변수명에 들어간 자료형 제거

- 피드백을 반영하여 자료형이 들어간 carNameList를 carNames로 변경!

Seongwon97 committed:
refactor(RacingGame.java): firstFarestPositon메서드에서

firstFarestPosition에서 stream결과 result의 size가 0일 경우 발생하는 error를 막을 코드 추가
as-is:
	private int findFarthestPosition() {
		return cars.stream()
			.sorted(Comparator.comparing(Car::getPosition))
			.collect(Collectors.toList())
			.get(cars.size() - 1).getPosition();
	}

to-be:
	private int findFarthestPosition() {
		List<Car> sortedCars = cars.stream()
				.sorted(Comparator.comparing(Car::getPosition).reversed())
				.collect(Collectors.toList());

		if (!sortedCars.isEmpty()) {
			return sortedCars.get(0).getPosition();
		}
		return 0;
	}

Seongwon97 committed:
refactor(RacingGame.java): joinWinnerNames메서드의 반환 문자열 생성 수정

- 피드백을 반영하여 String.join()에서 stream.collect(Collectors.joining())으로 수정
as-is:
	private String joinWinnerNames(final List<Car> winners) {
		final String[] winnerNames = winners.stream()
			.map(Car::getName)
			.toArray(String[]::new);
		return String.join(", ", winnerNames);
	}

to-be:
    private String joinWinnerNames(final List<Car> winners) {
        return winners.stream()
                .map(Car::getName)
                .collect(Collectors.joining(", "));
    }

Seongwon97 committed:
refactor(RacingGame.java): getWinners()메서드의 반환타입 변경

List값을 String으로 합쳐서 반환하는 joinWinnerNames()메서드의 작업을
getWinners()메서드에서 한번에 동작하도록 수정

as-is:
	private List<Car> getWinners(final int farthestPosition) {
		return cars.stream()
			.filter((car) -> farthestPosition == car.getPosition())
			.collect(Collectors.toList());
	}

	private String joinWinnerNames(final List<Car> winners) {
		final String[] winnerNames = winners.stream()
			.map(Car::getName)
			.toArray(String[]::new);
		return String.join(", ", winnerNames);
	}

to-be:
    private String getWinners(final int farthestPosition) {
        return cars.stream()
                .filter((car) -> farthestPosition == car.getPosition())
                .map(Car::getName)
                .collect(Collectors.joining(", "));
    }

Seongwon97 committed:
refactor(RacingGameMessage.java): 상수를 각 책임에 맞는 클래스로 이동

피드백을 반영하여 RacingGameMessage.java파일에 있는 상수들을 각 책임을 갖는 클래스로 이동

Seongwon97 committed:
refactor(Random.java): 불필요한 random의 생성 제거

static 으로 random객체를 생성하여 불필요하게 반복적인 random객체 생성을 제거하였다.

Seongwon97 committed:
fix(RandomNumber.java): 랜덤값 생성 범위를 요구사항에 맞게 수정

1~9사이의 수로 생성하던 랜덤값을 요구사항에 맞게 0~9사이의 수로 생성하도록 수정

as-is:
public class RandomNumber {

	private final static int MAX_EXCLUSIVE = 9;
	private static final Random random = new Random();

	public int getRandomNumber() {
		return random.nextInt(MAX_EXCLUSIVE) + 1;
	}
}

to-be:
public class RandomNumber {

	private final static int MAX_EXCLUSIVE = 10;
	private static final Random random = new Random();

	public int getRandomNumber() {
		return random.nextInt(MAX_EXCLUSIVE);
	}
}

Seongwon97 committed:
refactor(Input.java): 줄바꿈 명령을 lineSeparator()로 변경

줄바꿈 코드("\n")을 System.lineSeparator()로 변경

as-is:
...
catch (RuntimeException exception) {
   System.out.println(exception.getMessage() + "\n");
...

to-be:
...
catch (RuntimeException exception) {
   System.out.println(exception.getMessage() + System.lineSeparator());
...

Seongwon97 committed:
refactor(Output.java): 매개변수와 통일하여 메서드 명을 변경

pringWinner()메서드를 매개변수에 맞게 printWinners()로 변경

as-is:
public void printWinner(final String winners) {
...
}

to-be:
public void printWinners(final String winners) {
...
}

Seongwon97 committed:
refactor(/test): 테스트 클래스 들의 DisplayName을 통일

테스트 부분에 작성된 클래스들의 DisplayName의 형태를 통일화 하였다.

Seongwon97 committed:
refactor(Car.java): equals, hashCode, toString 재정의

- equals, hashCode을 재정의하여 객체 비교를 쉽게 할 수 있도록 수정
- toString을 재정의하여 객체 출력을 쉽게 할 수 있도록 수정

to-be:
	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		Car car = (Car) o;
		return Objects.equals(getName(), car.getName());
	}

	@Override
	public int hashCode() {
		return Objects.hash(getName());
	}

	@Override
	public String toString() {
		return name;
	}

Seongwon97 committed:
refactor(Input.java, InputValidation.java): 반복횟수가 음수일 떄 에러 매시지 추가

- 반복 횟수의 입력이 음수일 때 에러 메시지 추가
- Input.java파일에서 inputRepeats메서드의 반환 타입을 int로 변경하여 inputValidRepeats메서드 제거

Input.java파일
as-is:
	public int inputValidRepeats() {
		return Integer.parseInt(inputRepeats());
	}

	private String inputRepeats() {
		try {
			System.out.println(INPUT_REPEATS);
			final String repeats = scanner.next();
			inputValidation.validateRepeats(repeats);
			return repeats;
		} catch (RuntimeException exception) {
			System.out.println(exception.getMessage() + System.lineSeparator());
			return inputRepeats();
		}
	}

to-be:
	public int inputRepeats() {
		try {
			System.out.println(INPUT_REPEATS);
			final String repeats = scanner.next();
			int repeatsNumber = inputValidation.validateNotIntegerRepeats(repeats);
			inputValidation.validateNegativeNumberRepeats(repeatsNumber);
			return repeatsNumber;
		} catch (RuntimeException exception) {
			System.out.println(exception.getMessage() + System.lineSeparator());
			return inputRepeats();
		}
	}

InputValidation.java파일
as-is:
	public void validateRepeats(final String repeats) {
		if (!Pattern.matches(NUMBER_PATTERN, repeats)) {
			throw new RuntimeException(REPEATS_ERROR);
		}
	}

to-be:
	public int validateNotIntegerRepeats(final String repeats) {
		try {
			return Integer.parseInt(repeats);
		} catch (NumberFormatException exception) {
			throw new RuntimeException(REPEATS_NON_INTEGER_ERROR);
		}
	}

	public void validateNegativeNumberRepeats(final int repeatNumber) {
		if (repeatNumber < 0) {
			throw new RuntimeException(REPEATS_NEGATIVE_NUMBER_ERROR);
		}
	}

Seongwon97 committed:
refactor(InputValidationTest.java): 정상 실행 테스트 코드 수정

- 정상 실행되는 테스트 코드를 단순히 실행되는 것을 확인하는 verify에서
assertDoesNotThrow를 통한 예외가 발생하지 않는지 확인하는 코드로 변경
- 잘못된 반복 횟수 테스트의 input값 변경

checkRightName메서드
as-is:
		void checkRightName(final String names) {
			InputValidation validation = mock(InputValidation.class);
			validation.validateName(names);
			verify(validation).validateName(names);
		}

to-be:
		void checkRightName(final String names) {
			InputValidation validation = new InputValidation();
			assertDoesNotThrow(() -> validation.validateName(names));
		}

checkRightRepeats메서드
as-is:
		void checkRightRepeats(final String repeats) {
			InputValidation validation = mock(InputValidation.class);
			validation.validateNotIntegerRepeats(repeats);
			verify(validation).validateNotIntegerRepeats(repeats);
		}

to-be:
		void checkRightRepeats(final String repeats) {
			InputValidation validation = new InputValidation();
			assertDoesNotThrow(() -> validation.validateNotIntegerRepeats(repeats));
		}

Seongwon97 committed:
test(InputValidationTest.java): 음수의 반복 횟수 입력 테스트 추가

- InputValidation.java의 validateNegativeNumberRepeats메서드를
테스트하는 테스트 추가

Seongwon97 committed:
refactor(InputValidation.java): 반복적인 InputValidation객체 생성 제거

- InputValidation 객체 생성을 각각의 매서드가 아닌 클래스 내부에 하여
불필요한 InputValidation객체 생성을 제거

Seongwon97 commented:
미르님
보내주신 피드백 내용을 코드에 모두 반영해봤습니다.

피드백 반영을 하며 몇가지 궁금한 점이 생겨 질문을 드립니다.
1. [RacingGameMessage.java](https://github.com/woowacourse/java-racingcar/pull/326/files/0ca7dc3e1276abf5dd468d1a829ddb8dbfe0233f#diff-efd32d33e0af272bcb9b432db81b379d3abe58ceff21c574e0fe6b928d3f28f6) 파일의 피드백으로 "상수도 각 책임에 맞게 각 클래스에 정의하는 건 어떨까요?"라는 피드백을 주셨습니다. 
해당 내용은 프리코스를 진행할 때부터 메시지와 같은 상수를 한 곳에 모아서 보관을 해야할지 사용하는 클래스 내부에 보관을 해야할지 많은 고민을 했었습니다. 미르님께서는 각각의 클래스에 정의하는 것을 추천하는 것으로 이해했습니다. 하지만 제가 작성한 코드에서 Input.java파일과 InputValidation.java파일에서 같이 사용하는 CAR_NAME_DISTRIBUTOR같은 것은 중복된 내용의 상수인데 이러한 경우에는 한 쪽에서 public으로 정의를 하고 다른 한쪽에서 Import를 사용하는 것이 좋을지 아니면 양쪽에 private으로 각각 정의를 하여 사용하는 것이 좋을지, 아니면 이러한 상수만 따로 클래스를 생성하여 관리하는 것이 좋을지 궁금합니다. 실무에서는 주로 어떠한 방법을 사용하는지 궁금합니다!!🧐

2. [CarTest.java](https://github.com/woowacourse/java-racingcar/pull/326/files/0ca7dc3e1276abf5dd468d1a829ddb8dbfe0233f#diff-e64fd1c2c98dd6db6e7766846ef647c0d831d791ccf9d3a8097b42c4a31c1a4e) 파일의 피드백으로 주신 "해당 내용은 Mock으로 구현하는 것도 좋지만 다른 방식으로 구현해보는 건 어떨까요?"의 피드백을 받고 Mock이외에 `goForward`메서드를 테스트 할 방법을 생각해봤습니다.
하지만 해당 테스트는 아무리 생각을 해봐도 Mock으로 테스트하는 방법 이외에 어떠한 방법으로 테스트를 진행해야할지 떠오르지가 않네요...
goForward()메서드가 RandomNumber객체의 getRandomNumber()메서드의 결과값에 따라 전진, 유지의 결과가 나와서 Mock이외에 테스트를 할 방법이 떠오르지 않습니다 🥲
혹시 다른 테스트 방법이 있다면 알려주실 수 있으실까요..?

감사합니다. 🙇🏻‍♂️

ddu0422 commented on src/main/java/racingcar/controller/RacingGame.java:
```diff
@@ -0,0 +1,74 @@
+
+    private final Input input;
+    private final Output output;
+    private List<Car> cars;
```
해당 값도 final이여도 괜찮지 않을까요 !_!

ddu0422 commented on src/main/java/racingcar/models/Car.java:
```diff
@@ -0,0 +1,53 @@
+	private static final int FORWARD_STANDARD = 4;
+	private static final int INITIAL_POSITION = 0;
+
+	final RandomNumber randomNumber;
+	final String name;
+	int position;
```
해당 값들도 내부에서만 사용하고 있는데 package-private으로 선언한 특별한 이유가 있을까요?

ddu0422 commented on README.md:
```diff
@@ -2,6 +2,24 @@
-## 우아한테크코스 코드리뷰
+# ToDo List
+- 난수 생성 기능 구현
+  - test: 1~9사이의 난수가 생성되었는지 확인
```
요구사항 내용도수정되어야겠군요 !

ddu0422 reviewed (approved):
안녕하세요 렉스 😄 

리뷰 반영해주셔서 감사해요. 
리뷰는 제가 이야기한 방향으로 전부 반영하지 않으셔도 괜찮아요. 렉스와 생각이 다른 부부이 있다면 언제든지 말씀해주셔도 괜찮아요 😉 

많은 고민을 한 흔적들이 보여요. 앞으로 그런 고민들을 계속하신다면 강의와 피드백을 받지 않아도 스스로 발전하는 개발자가 될 수 있을거예요 😄 앞으로도 화이팅입니다 ㅎ_ㅎ

해당 pr은 머지할게요 !_! 첫 미션 수고하셨습니다!!

---

Q&A는 해당 comment에 답글 남겼어요. 보기 편하시게 resolve 해제 해놨습니다 :)
추가로 드리는 리뷰와 함께 고민하며 다음 미션에 반영해도 괜찮을 것 같아요.

깃 커밋 메시지 부분도 고민해보면 좋을 것 같아요. 해당 부분에서 많은 시간을 쓰시는 것 같아서요 😭
as-is /  to-be 부분은 작성하지 않으셔도 될 것 같아요. 깃을 공부하시다보면 아시겠지만, 깃으로 as-is와 to-be를 확인할 수 있습니다 !_!
